src/Microsoft.DotNet.CodeFormatting/Analyzers/UnwrittenWritableFieldAnalyzer.cs

```diff
@@ -0,0 +1,351 @@
+                        {
+                            MarkFieldAsWritten(fieldSymbol);
+                        }
+                    }
+                }
+            }
+        }
+
+        private bool IsImmutablePrimitiveType(ITypeSymbol type)
+        {
+            // All of the "special type" structs exposed are all immutable,
+            // so it's safe to assume all methods on them are non-mutating, and
+            // therefore safe to call on a readonly field
+            return type.SpecialType != SpecialType.None && type.TypeKind == TypeKind.Struct;
+        }
+
+        private void ReportUnwrittenFields(CompilationAnalysisContext context)
+        {
+            IEnumerable<IFieldSymbol> fieldsToMark = _candidateReadonlyFields.Except(_writtenFields);
+            foreach (var field in fieldsToMark)
+            {
+                context.ReportDiagnostic(Diagnostic.Create(rule, field.Locations[0], field.Name));
+            }
+        }
+
+        private void CheckForFieldWrite(ExpressionSyntax node, SemanticModel model)
+        {
+            var fieldSymbol = model.GetSymbolInfo(node).Symbol as IFieldSymbol;
+
+            if (fieldSymbol != null)
+            {
+                if (IsInsideOwnConstructor(node, fieldSymbol.ContainingType, fieldSymbol.IsStatic, model))
+                {
+                    return;
+                }
+
+                MarkFieldAsWritten(fieldSymbol);
+            }
+        }
+
+        private void MarkFieldAsWritten(IFieldSymbol field)
+        {
+            _writtenFields.Add(field);
+        }
+
+        private bool IsInsideOwnConstructor(SyntaxNode node, ITypeSymbol type, bool isStatic, SemanticModel model)
+        {
+            while (node != null)
+            {
+                switch (node.Kind())
+                {
+                    case SyntaxKind.ConstructorDeclaration:
+                        {
+                            return model.GetDeclaredSymbol(node).IsStatic == isStatic &&
+                                IsInType(node.Parent, type, model);
+                        }
+                    case SyntaxKind.ParenthesizedLambdaExpression:
+                    case SyntaxKind.SimpleLambdaExpression:
+                    case SyntaxKind.AnonymousMethodExpression:
+                        return false;
+                }
+
+                node = node.Parent;
+            }
+
+            return false;
+        }
+
+        private bool IsInType(SyntaxNode node, ITypeSymbol containingType, SemanticModel model)
+        {
+            while (node != null)
+            {
+                if (node.IsKind(SyntaxKind.ClassDeclaration) || node.IsKind(SyntaxKind.StructDeclaration))
+                {
+                    return Equals(containingType, model.GetDeclaredSymbol(node));
+                }
+
+                node = node.Parent;
+            }
+
+            return false;
+        }
+    }
+
+    internal static class FieldSymbolExtensions
```

Comment:
I don't know yet that they are generally useful. These were originally private methods within the rule class, like `IsFieldSerializableByAttribute(IFieldSymbol)`, which I re-cast as extension methods because I like being able to write `if (symbol.IsSerializableByAttributes())`. If they come in useful elsewhere I'll extract them.